app/javascript/extensible-components/index.ts

```diff
@@ -0,0 +1,69 @@
+import { IExtensionComponent, IMiQApiCallback } from './lib';
+
+const source = new Rx.Subject();
+const items = new Set();
+
+/**
+ * Class for easy creation of extensible component.
+ */
+export class ExtensibleComponent {
+  public unsubscribe: () => void;
+  constructor(public name: string, public api: IMiQApiCallback, public render: IMiQApiCallback){}
+}
+
+/**
+ * Helper function to create new component.
+ * @param name string name of new component.
+ * @param api callback functions to change inner logic of component.
+ * @param render callback function to apply render functions.
+ */
+function addComponent(name: string, api?: IMiQApiCallback, render?: IMiQApiCallback) {
+  let newCmp = new ExtensibleComponent(name, api, render);
+  source.onNext({action: 'add', payload: newCmp});
+  return newCmp;
+}
+
+/**
+ * Helper function to subscribe to extension items based on component's name.
+ * @param cmpName name of component we want to subscribe to.
+ * @return object which has with and unsubscribe property. With is for callback to use found component and delete to
+ * unsubscribe from rxjs subject.
+ */
+function subscribe(cmpName: string) {
+  let unsubscribe;
+  return {
+    with: (callback) => {
+      unsubscribe = source
+        .map(sourceAction => sourceAction.action === 'add' ? sourceAction.payload : {})
+        .filter(component => component && component.name === cmpName)
+        .subscribe(cmp => cmp && callback(cmp));
+
+      ManageIQ.extensions.items.forEach((component) => {
+        component.name === cmpName && callback(component);
+      });
+    },
+    delete: () => unsubscribe && unsubscribe()
+  }
+}
+
+const extensions = {
+  addComponent: addComponent,
```

Comment:
You can use the object literal shorthand form 😄 e.g. just `addComponent` instead of `addComponent: addComponent`.

Reply:
consider adding a tool like lebab to your editor to auto convert those.

Reply:
Nice catch! Looking at [Lebab](https://github.com/lebab/lebab), it's the opposite of Babel and using it as editor plugin seems most reasonable to me.